File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/UrlSafetyTest.kt
package com.flare.im.ui

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * DoD 32 — the vector table of spec/security-boundary.json. Every hostile entry
 * is something an attacker can put in a message; the four platforms run the same
 * ids and tooling/check-security-boundary.mjs fails when one stops.
 */
class UrlSafetyTest {
    private val hostile = listOf(
        "javascript.plain" to "javascript:alert(1)",
        "javascript.uppercase" to "JaVaScRiPt:alert(1)",
        "javascript.leadingSpace" to "   javascript:alert(1)",
        "javascript.embeddedTab" to "java\tscript:alert(1)",
        "javascript.embeddedNewline" to "java\nscript:alert(1)",
        "data.html" to "data:text/html,<script>alert(1)</script>",
        "data.base64" to "data:text/html;base64,PHNjcmlwdD5hbGVydCgxKTwvc2NyaXB0Pg==",
        "vbscript" to "vbscript:msgbox(1)",
        "file" to "file:///etc/passwd",
        "blob" to "blob:https://evil.example/9b2d",
        "about" to "about:blank",
        "empty" to "",
        "whitespace" to "   ",
        "notAUrl" to "just some text",
    )

    private val allowed = listOf(
        "http" to "http://example.com/path?q=1",
        "https" to "https://example.com/path#anchor",
        "schemeless" to "example.com/path",
        "schemelessWithPort" to "example.com:8443/path",
    )

    @Test fun refusesEveryHostileVector() {
        for ((id, input) in hostile) {
            assertNull(safeExternalUrl(input), id)
            assertFalse(isSafeExternalUrl(input), id)
        }
    }

    @Test fun allowsPlainWebAddresses() {
        for ((id, input) in allowed) assertTrue(isSafeExternalUrl(input), id)
    }

    @Test fun readsASchemeLessStringAsHttpsAndNeverRescuesOneThatHasAScheme() {
        assertEquals("https://example.com", safeExternalUrl("example.com"))
        assertNull(safeExternalUrl("javascript:alert(1)"))
    }

    @Test fun refusesNullWithoutThrowing() {
        assertNull(safeExternalUrl(null))
        assertFalse(isSafeExternalUrl(null))
    }
}

File: packages/flutter-im-ui/test/url_safety_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flare_im_ui/flare_im_ui.dart';

// DoD 32 —— spec/security-boundary.json 的向量表。每条恶意向量都是攻击者能塞进
// 消息里的字符串;四端跑同一批 id,tooling/check-security-boundary.mjs 会在某一端
// 不再验某条时报红。
void main() {
  const hostile = <String, String>{
    'javascript.plain': 'javascript:alert(1)',
    'javascript.uppercase': 'JaVaScRiPt:alert(1)',
    'javascript.leadingSpace': '   javascript:alert(1)',
    'javascript.embeddedTab': 'java\tscript:alert(1)',
    'javascript.embeddedNewline': 'java\nscript:alert(1)',
    'data.html': 'data:text/html,<script>alert(1)</script>',
    'data.base64':
        'data:text/html;base64,PHNjcmlwdD5hbGVydCgxKTwvc2NyaXB0Pg==',
    'vbscript': 'vbscript:msgbox(1)',
    'file': 'file:///etc/passwd',
    'blob': 'blob:https://evil.example/9b2d',
    'about': 'about:blank',
    'empty': '',
    'whitespace': '   ',
    'notAUrl': 'just some text',
  };

  const allowed = <String, String>{
    'http': 'http://example.com/path?q=1',
    'https': 'https://example.com/path#anchor',
    'schemeless': 'example.com/path',
    'schemelessWithPort': 'example.com:8443/path',
  };

  test('每条恶意向量都被拒', () {
    hostile.forEach((id, input) {
      expect(safeExternalUrl(input), isNull, reason: id);
      expect(isSafeExternalUrl(input), isFalse, reason: id);
    });
  });

  test('普通网址放行', () {
    allowed.forEach((id, input) {
      expect(isSafeExternalUrl(input), isTrue, reason: id);
    });
  });

  test('不带 scheme 的按 https 读,带 scheme 的绝不被救', () {
    expect(safeExternalUrl('example.com'), 'https://example.com');
    expect(safeExternalUrl('javascript:alert(1)'), isNull);
  });

  test('null 不抛异常', () {
    expect(safeExternalUrl(null), isNull);
    expect(isSafeExternalUrl(null), isFalse);
  });
}
